### DISCUSSION TITLE
Allow adding "Optional System Fields" post collection creation

### DISCUSSION THREAD

**Discussion by dstoyanoff** [directus/directus#8458] [Feature Requests]
This is a suggestion that came up after a discussion with @benhaynes and @rijkvanzanten in Discord.

Basically, at the moment, there are "shortcuts" for commonly used fields that the user can choose upon creating a new collection. They are called "Optional System Fields". 

If you however do not choose them while creating the collection, you cannot use those shortcuts later, thus if you need one, you need to create the field manually.

As @benhaynes suggested, this might instead be completely removed from the "Create Collection" workflow and instead added to the "Create Field" workflow.

**Comment by letoast** [comment]
Maybe in it's own "System fields" section in the dropdown?

**Comment by TomS-** [comment]
To add, 
"created_by" doesn't seem to be able to be added at a later date, even manually. At least through the UI.

**Reply by dstoyanoff** [reply]
Yes, this is the problem in this case. Once the collection was created, you cannot leverage those "shortcuts" to create such fields and you have to create the field manually just like any other field.

**Reply by benhaynes** [reply]
As a workaround in the meantime, you could go to a different collection that has one of those helper fields, and duplicate it to the new collection.

**Reply by Blanco237** [reply]
This doesn't work for the user_created and user_updated fields

**Comment by simonking76** [comment]
This is a bigger issue when using Directus on an existing database. For pre-existing tables you start to manage with Directus, you cannot add system fields, such as status, sort, created by etc.

I recall it was possible to add system fields to existing tables pre v9 so seems a shame it can't be 'easily' done now, especially a big selling point of directus is being able to manage existing databases, which seems to have just gotten a lot harder in v9

**Reply by rijkvanzanten** [reply]
> For pre-existing tables you start to manage with Directus, you cannot add system fields, such as status, sort, created by etc.

You totally can, there's no magic that happens with any of those fields. The shortcuts in the "create a collection" drawer are literally just that: a shortcut. There's nothing stopping you from adding a sort field to any other collection after the fact 🙂 

> I recall it was possible to add system fields to existing tables pre v9 so seems a shame it can't be 'easily' done now, especially a big selling point of directus is being able to manage existing databases, which seems to have just gotten a lot harder in v9

This is because in v8, "system fields" were actually special fields that were hardcoded to do certain things. That's no longer the case in v9 👍🏻

**Reply by TomS-** [reply]
The only field I can think of where some magic does happen is `created_by` and `user_updated`, I don't believe you can create these through the UI after the collection is created.

**Reply by benhaynes** [reply]
Do you know what the limiting factor is with creating these after the fact?

**Reply by simonking76** [reply]
These are the system fields, with notes on limiting factors:

sort
I can add a field called 'sort' to a pre-existing table, but I cannot get it to behave like a sort field as one does behave when creating a new table in directus with a system sort field.

The sort behaviour in the list view of collection does not kick in, where I click the sort header to then activate being able to drag n drop to change the order of items.

user_created 
user_updated
I can add these fields on existing tables as many to one fields, related to directus_users table. But I'd have to select a user when creating an item, or select the user who is doing the change when updating a item. It doesn't behave like the system fields where the user_updated field is auto-populated with who edited the item last - as a system user_updated field would.

date_created
date_updated
again these would have to be populated after being added to a pre-existing table, they are not auto-populated after being added as a datetime field, although with a bit more effort I guess I can maybe set these to be auto-populated with the current datetime. Not sure how that works when editing an existing record that already has a date_updated to then override it and change it to the current datetime on save... but seems a lot more work that just creating a 'system' date_updated field that does it all in the background.

status
this one is ok I can just create a status field with the statuses I need as per the system field.

**Reply by dstoyanoff** [reply]
As for the sorting, are you then selecting that field in the collection configuration? 
![image](https://user-images.githubusercontent.com/5079334/141275086-915df76e-81c8-44e7-880e-668028d5e485.png)

**Reply by rijkvanzanten** [reply]
> The only field I can think of where some magic does happen is created_by and user_updated, I don't believe you can create these through the UI after the collection is created.

Nope! The only thing it does is pre-fill the "On Create" option for a generic M2O UUID field:

<img width="611" alt="CleanShot 2021-11-11 at 15 33 42@2x" src="https://user-images.githubusercontent.com/9141017/141365166-c2228631-81d4-44c7-84c3-7bdbed4ab2e3.png">


As for @simonking76's list:

* sort — Any integer field can be used, is configured under the "Sort Field" option on the collection detail page, as outlined by @dstoyanoff
* user_created / user_updated — Same as mentioned above, is controlled through the "On Create" option under Schema of the field setup flow
* date_created / date_updated — Similar to the user ones, this can be controlled under the "On Create" and "On Update" options under the Schema tab of the (advanced) field setup flow
* status — Like you said, it's just any old dropdown 🙂

**Reply by kevincam3** [reply]
It would definitey be nice to just have these available to select like all the other options as opposed to having to create, copy, etc to get them. Can these be done as an interface? Is there a way to see an example/template of how to go about doing this?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by JohannesPertl** [comment]
Any updates on this one?

**Comment by on3dd** [comment]
Is there any news on this? It's been almost 2 years since the last message

**Reply by rijkvanzanten** [reply]
Low priority as any of those 'system fields' can just be added as fields. The system fields setup at the beginning is just a shortcut, but not the only place to configure this 👍🏻

**Reply by on3dd** [reply]
Got it, thank you mate!

**Comment by intensr** [comment]
+1 cool feature
